rep-ixxxx.rst

```diff
@@ -0,0 +1,826 @@
+
+
+Endianness
+----------
+
+In order to accomodate controllers that need it, the generic nodes in ``industrial_robot_client`` support both little and big-endian data streams.
+These are implemented as *swap* and *non-swap* variants of all nodes in the package.
+
+As an example: a little-endian client wishing to connect to a server running on a controller that uses network (or big-endian) byte ordering should use the *swap* version of the client nodes.
+A little-endian client connecting to a controller sending data in little-endian byte order should use the *non-swapping* version of the client nodes (as byte swaps are not needed in this case).
+
+The following table shows the supported variants (type sizes are in bytes)::
+
+              Type
+  Node type   Integer   Real
+
+  No swap           4      4
+  Swap              4      4
+  No swap           4      8
+
+Swapping node variants will byte-swap fields of incoming messages, while taking field sizes into account.
+
+
+Message Layout
+==============
+
+The following sections describe the different sub structures that make up a valid Simple Message message.
+
+
+Prefix
+------
+
+All messages must start with the *prefix*, which may only contain a single field: ``length``.
+Message length is defined as the sum in bytes of the sizes of the individual fields in the *header* and the *body*, excluding the ``length`` field itself (ie: only actual message bytes are considered).
+
+Layout::
+
+  length           : shared_int
+
+Notes
+
+#. Refer to section `Shared Types`_ for information on the size of supported field types.
+#. The size of fields that are arrays or lists shall be defined as the size of their base type (ie: ``shared_int``) multiplied by the number of elements in the list, or the declared size of the array.
+   Example: an array of ``shared_int`` with ten (``10``) elements in it has a total size of forty (``40``) bytes.
+
+
+Header
+------
+
+The next three fields make up the *message header*, which is that part of the message that encodes the message type (or it's intent), whether the message is a topic-like broadcast, a service request or reply and, if it is a reply, what the result of the service request was.
+
+Layout::
+
+  msg_type         : shared_int
+  comm_type        : shared_int
+  reply_code       : shared_int
+
+Notes
+
+#. Refer to [#REP-I0004]_ for valid values for the ``msg_type`` field.
+#. Refer to `Communication Types`_ for valid values for the ``comm_type`` field.
+#. Refer to `Reply Codes`_ for valid values for the ``reply_code`` field.
+#. For ``TOPIC`` and ``SERVICE_REQUEST`` type messages, the ``reply_code`` field must be set to ``INVALID``.
+#. The ``SUCCESS`` and ``FAILURE`` reply codes shall only be used with ``SERVICE_REPLY`` type messages.
```

Comment:
This part is a bit ambiguous.  In implementations, I've seen a `FAILURE` behave in multiple ways:
 - indicate the request failed AND no `DATA` payload
 - indicate the request failed AND the `DATA` payload although present, can be ignored.
 - indicate the request failed AND the `DATA` payload contains failure information
 - indicate the request failed AND the `DATA` payload contains a special payload with failure information

To remove all doubt, this should follow the ROS convention, but I'm not entirely sure what the ROS convention is.  A service call returns a boolean, which could be thought of as the `reply_code`.  However, in the event a ROS service returns `false`, I don't believe the reply is populated (even if the service server populated it).

Reply:
> However, in the event a ROS service returns `false`, I don't believe the reply is populated (even if the service server populated it).

The reply *may* be populated / initialised, but typically only just enough for the error / failure to be communicated (ie: setting an `error_message` or `return_code` field).

As to the four variants you mention: 1 is the most efficient, and I've been using that personally, but only if the semantics of returning a 'header only' message with a `FAILURE` are clear.

`motoman_driver` uses option 3 (or 4, not sure what the difference between those two is).

But is the text really ambiguous? The section describes the `Header` and the acceptable values for each field. Not necessarily how those are used for a particular message.

Reply:
I'm also slightly confused by your comment.

> A service call returns a boolean, which could be thought of as the `reply_code`. 

to me, `reply_code` is exactly the `bool` you get back when invocation of a ROS service fails.

> However, in the event a ROS service returns `false`, I don't believe the reply is populated (even if the service server populated it).

Ok. So what do you suggest we spec here? The text currently just states that the reply codes `FAILURE` and `SUCCESS` may only be used with messages of type `SERVICE_REPLY`. Nothing more.

Would you like to see a section detailing the semantics of the various ways in which a failure / error can actually be represented / transmitted?

Reply:
Let's just agree that it should mirror a ROS service.  I agree that the `reply_code` should match exactly the ROS Service boolean return.

However, in the event of a failure, the ROS service does not populate the `response`, see [here](http://wiki.ros.org/roscpp/Overview/Services#Callback_Signature), specifically:
>A return value of true means the service succeeded, and the response object has been filled with the necessary data. A return value of false means the call has failed and the response object will not be sent to the caller.

To be consistent, a `reply_code` of `FAILURE` means that the data payload will be empty.  If failure data is required (i.e. an error code), then the `reply_code` will be set to `SUCCESS` and it's up to the caller to parse the response.

This is not the most intuitive, but I feel consistency is better than nothing.

Reply:
I guess I'm advocating for a fixed service reply/response message (like ROS).  Specifically "2. indicate the request failed AND the DATA payload although present, can/should be ignored."

Reply:
For future readers: addressed this in 372be0663b92ee914d64e0a1ddc0af9e512bbdf1.